registry/rpc/rpcregistry.go

```diff
@@ -34,12 +34,31 @@ type RPCRegistry struct {
 	mu             *sync.Mutex
 	registryClient pb.RegistryClient
 	registryConn   *grpc.ClientConn
+
+	cachesMu       *sync.Mutex
+	heartbeatCache map[string]heartbeatCacheLine
+	unitStateCache map[string]unitStateCacheLine
+}
+
+type heartbeatCacheLine struct {
+	unitName string
+	machID   string
+	deadline time.Time
+}
+
+type unitStateCacheLine struct {
+	unitName  string
+	unitState *unit.UnitState
+	deadline  time.Time
 }
 
 func NewRPCRegistry(dialer func(string, time.Duration) (net.Conn, error)) *RPCRegistry {
 	return &RPCRegistry{
-		mu:     new(sync.Mutex),
-		dialer: dialer,
+		mu:             new(sync.Mutex),
```

Comment:
This won't work with the current version as we always create a new rpc registry to connect to a new engine

For example:
```
r.rpcRegistry = NewRPCRegistry(r.rpcDialer)
				// connect to rpc registry
				r.rpcRegistry.Connect()
```